lua/tardis/libraries/libraries/sh_debug_portals.lua

```diff
@@ -0,0 +1,316 @@
+            epo_z = val
+            UpdatePortalExitOffset()
+        end)
+
+        local eaop, eaop2 = SetupProperty( "Exit Point", "Pitch", eao_p, 360, function(val)
+            eao_p = val
+            UpdatePortalExitOffset()
+        end)
+        local eaoy, eaoy2 = SetupProperty( "Exit Point", "Yaw", eao_y, 360, function(val)
+            eao_y = val
+            UpdatePortalExitOffset()
+        end)
+        local eaor, eaor2 = SetupProperty( "Exit Point", "Roll", eao_r, 360, function(val)
+            eao_r = val
+            UpdatePortalExitOffset()
+        end)
+
+
+        local ep_category = pr:GetCategory("Exit Point")
+
+        ep_category.Container:SetVisible(false)
+        ep_category.Expand:SetExpanded(false)
+        ep_category:InvalidateLayout()
+
+        local reset = pr:CreateRow( "Actions", "Reset" )
+        reset:Setup( "Bool" )
+        reset:SetValue(false)
+        reset.DataChanged = function( _, val )
+            reset:SetValue(false)
+
+            px, py, pz = orig_px, orig_py, orig_pz
+            ang_p, ang_y, ang_r = orig_ang_p, orig_ang_y, orig_ang_r
+            thickness = orig_thickness
+            inverted = orig_inverted
+            width = orig_width
+            height = orig_height
+
+            epo_x, epo_y, epo_z = orig_epo_x, orig_epo_y, orig_epo_z
+            eao_p, eao_y, eao_r = orig_eao_p, orig_eao_y, orig_eao_r
+
+            UpdatePortalPos()
+            UpdatePortalAng()
+            UpdatePortalSize()
+            UpdatePortal3D()
+            UpdatePortalExitOffset()
+
+            frame:SetDeleteOnClose(true)
+            frame:Close()
+            frame:Remove()
+        end
+
+        local inv = pr:CreateRow( "Actions", "Print to console" )
+        inv:Setup( "Bool" )
+        inv:SetValue(false)
+        inv.DataChanged = function( _, val )
+            inv:SetValue(false)
+
+            print("Portal = {")
+            print("\t-- Generated by portals debug tool")
+            print("\tpos = Vector(" .. px .. ", " .. py .. ", " .. pz .. "),")
+            print("\tang = Angle(" .. ang_p .. ", " .. ang_y .. ", " .. ang_r .. "),")
+            print("\twidth = " .. width .. ",")
+            print("\theight = " .. height .. ",")
+
+            if thickness and thickness ~= 0 then
+                print("\tthickness = " .. thickness .. ",")
+            end
+            if inverted then
+                print("\tinverted = " .. tostring(inverted) .. ",")
+            end
+
+            print("),")
+        end
+    end
+
+    net.Receive("TARDIS-Debug-Portals", function()
+        local p = net.ReadEntity()
+        TARDIS:ShowPortalDebugMenu(p)
+    end)
+end
+
+concommand.Add("tardis2_debug_portals", function(ply,cmd,args)
+    if not ply:IsAdmin() then return end
+
+    local portal = wp.GetFirstPortalHit(ply:EyePos(), ply:EyeAngles():Forward())
+
+    if IsValid(portal.Entity) then
```

Comment:
Is it worth printing a message to the user if they are not pointing at a portal?

Reply:
Done